toolchain/parser/parser_handle_index_expression.cpp

```diff
@@ -0,0 +1,31 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/parser/parser_context.h"
+
+namespace Carbon {
+
+auto ParserHandleIndexExpression(ParserContext& context) -> void {
+  auto state = context.PopState();
+  state.state = ParserState::IndexExpressionFinish;
+  context.PushState(state);
+  context.AddNode(ParseNodeKind::IndexExpressionStart, context.Consume(),
+                  state.subtree_start, state.has_error);
+  if (context.PositionIs(TokenKind::CloseSquareBracket)) {
+    CARBON_DIAGNOSTIC(ExpectedExpressionAfter, Error,
+                      "Expected integer literal.");
```

Comment:
While literals might be the only thing that you support right now for tuple literals, I think we should aim to have this message be more generic for index expressions in general, where expressions may generally be used. Something like "Expected an expression." would be better. Similarly I'd suggest just "ExpectedExpression" for the name since you don't have any "after" component to the diagnostic.

Reply:
You've written "Expected expression inside the square bracket.", but note this is grammatically correct ("inside the square bracket" is singular, there are two brackets). Also, "ExpectedExpressionInside" is fairly generic relative to the specific nature of the message.

That said, I think this is something we can improve later. Change it to just "Expected an expression." and ExpectedExpression for the DiagnosticKind. Note here: (a) the message already provides the location where an expression was expected, and (b) although ExpectedExpression is generic, so is the message.